### ISSUE TITLE
SVG Coloring takes too much time in GPU thread

### ISSUE THREAD

**Issue by FrantisekGazo** [dnfield/flutter_svg#447]
When I apply a color to the `SvgPicture` there's too much work on the GPU thread.

**no color:**
<img width="503" alt="Screenshot 2020-11-10 at 11 10 26" src="https://user-images.githubusercontent.com/5168594/98661290-3b71d000-2381-11eb-9442-ae74dcfc1f16.png">
**with color:**
<img width="501" alt="Screenshot 2020-11-10 at 11 10 16" src="https://user-images.githubusercontent.com/5168594/98661297-3dd42a00-2381-11eb-8696-010cfab092bc.png">

### Steps to reproduce:
- I created a simple example project: https://github.com/FrantisekGazo/flutter-svg-issue-example
- Keep or remove the `color: Colors.red,` line.
- Run a profile build and turn on the performance overlay.
- Scroll up and down in the demo app.

**Comment by dnfield** [comment]
Yup. Color filtering requires a saveLayer, which is expensive. If this is too expensive for your picture/device, you should consider creating the colors you need _without_ filtering.

**Comment by dnfield** [comment]
Closing this as WAI - there's not really much that can be done here, since this is just a relatively expensive effect.

**Comment by FrantisekGazo** [comment]
@dnfield And is it somehow possible to change the "fill" color in svg from code instead of using a `ColorFiltered` widget?

**Comment by FrantisekGazo** [comment]
I now created this class for that purpose and it seems to work fine.
```dart
///
/// Provides a SVG image that can be colored by changing it's fill color.
///
class ColoredExactAssetPicture extends ExactAssetPicture {
  static String _name(String asset, Color color) => '$asset-${color?.value}';

  static PictureInfoDecoder<String> _coloredDecoder(Color color) {
    return (data, _, key) async {
      final baseRoot = await svg.fromSvgString(data, key);
      final root = (color != null)
          ? baseRoot.mergeStyle(
              DrawableStyle(
                fill: DrawablePaint(PaintingStyle.fill, color: color),
              ),
            )
          : baseRoot;
      return PictureInfo(
        picture: root.toPicture(
          clipToViewBox: false,
          size: root.viewport.viewBox,
        ),
        viewport: root.viewport.viewBoxRect,
        size: root.viewport.size,
      );
    };
  }

  /// This key is used to cache colored images separately.
  final String _coloredAssetName;

  /// This is the original key matching a svg asset.
  final String _assetName;

  ColoredExactAssetPicture(this._assetName, Color color)
      : this._coloredAssetName = _name(_assetName, color),
        super(_coloredDecoder(color), _name(_assetName, color));

  @override
  PictureStreamCompleter load(
    AssetBundlePictureKey key, {
    PictureErrorListener onError,
  }) {
    // replace the key so the the real svg asset is found
    final realKey = AssetBundlePictureKey(
      bundle: key.bundle,
      name: key.name.replaceAll(_coloredAssetName, _assetName),
    );
    return super.load(realKey, onError: onError);
  }
}
```
And usage:
```dart
SvgPicture(
   ColoredExactAssetPicture('assets/icon/example.svg', Colors.red),
),
```

@dnfield Can you recommend a solution that would be cleaner?

**Comment by dnfield** [comment]
That won't work if different opacities are used with overlapping colors.

But for some cases it should be fine.

**Comment by dnfield** [comment]
Hmm.

I'm now wondering if this regressed compared to changes from 0.18.0. Checking on that.

**Comment by dnfield** [comment]
I'm not able to see significant differences between 18.x and 19.x :\ I'm also not able to see raster times as bad as what you're posting from your sample app, eithe ron a macOS app or a Pixel 4.

**Comment by FrantisekGazo** [comment]
On my pixel 2 it seems fine too. But we test on some lower/mid end devices at work and there I saw those big differences.

But in the end I managed to fix those issues with a modified version of the `ColoredExactAssetPicture` class I posted above. The `mergeStyle()` was causing some issues and handling those resulted in an unnecessarily complicated solution. So instead of using `mergeStyle()` I simply call `String.replaceAll()` on loaded String `data` and replace stroke and fill colors.
